### PR TITLE
Reverse engineer SQL Server views

### PR DESCRIPTION
**Pull request by ErikEJ**
part of ~#1697~ #1679

### CODE REVIEW

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs [hunk 1]
```diff
@@ -545,6 +601,31 @@ private static string EscapeLiteral(string s)
...
+    [tp].[name] AS [type_name],
+    CAST([c].[max_length] AS int) AS [max_length],
+    CAST([c].[precision] AS int) AS [precision],
+    CAST([c].[scale] AS int) AS [scale],
+    [c].[is_nullable],
+    [c].[is_identity],
+    [dc].[definition] AS [default_sql],
+    [cc].[definition] AS [computed_sql]
+FROM [sys].[columns] AS [c]
+JOIN [sys].[views] AS [v] ON [c].[object_id] = [v].[object_id]
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Can we run single command to get this data?
Like doing Union on sys.views & sys.tables before performing join.

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs [hunk 2]
```diff
@@ -507,18 +509,72 @@ private static string EscapeLiteral(string s)
...
+            DbConnection connection,
+            Func<string, string, string> tableFilter,
+            IReadOnlyDictionary<string, (string, string)> typeAliases)
+        {
+            using (var command = connection.CreateCommand())
+            {
+                var views = new List<DatabaseTable>();
+
+                var commandText = @"
+SELECT
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Can we combine this one too with other using Union? This one may be bit more tricky than columns one due to different filters but there is scope for DRY.

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 1 above)
Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Will that make the code more readable and maintainable?

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 2 above)
Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Will that improve readability and maintainability?

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: From my understanding there is only 1 line difference between both SQL. Also table/views both are same in EF model so it making them same on database end. 0 issue in readability and maintainability here.

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Makes code DRY. For this case direct UNION on top level is good enough. Gives single command and only one code path to read data.

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs (hunk 1 above)
Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: I will give it a shot

**Comment by ErikEJ** [comment]
Updated

File: src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs [hunk 3]
```diff
@@ -476,9 +477,39 @@ private static string EscapeLiteral(string s)
...
+                commandText = commandText + @"
 WHERE " + filter;
 
+                var viewCommandText = @"
+UNION
+SELECT
+    SCHEMA_NAME([v].[schema_id]) AS [schema],
+    [v].[name]";
+
+                if (supportsMemoryOptimizedTable)
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: Does view has memory optimized?

Comment (ErikEJ) [reply src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs]: No, but the 2 queries must have the same shape

**Pull request merged**